packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
-      const { data } = message;
-      const { confirmations } = data;
+    it("should throw when subscribing with invalid trigger", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("invalid", () => {})).to.be.rejected;
+    });
 
-      if (confirmations === 7) {
-        subscription.removeAllListeners();
+    it("should throw when subscribing without callback", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      return expect(subscription.on("confirmation")).to.be.rejected;
+    });
+
+    it("should call a write contract method (send tx) - imediate subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+
+          // FIXME: UnhandledPromiseRejectionWarning
+          //expect(1).to.equal(2);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should call a write contract method (send tx) - late subscription", async () => {
+      var rand = Math.floor(Math.random() * Math.floor(1000)).toString();
+      subscription = simpleStorage.setValue(rand);
+      const fakeFn = sinon.fake();
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
+
+      const onMessage = async message => {
+        const { data } = message;
+        const { confirmations } = data;
+
+        if (confirmations >= 7) {
+          fakeFn();
+
+          subscription.removeAllListeners();
+
+          const value = await simpleStorage.getValue();
+          expect(value).to.equal(rand);
+        }
+      };
+
+      await subscription.on("confirmation", onMessage);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(fakeFn.called).to.be.true;
+    });
+
+    it("should remove listener after timeout", async () => {
+      subscription = simpleStorage.setValue("hello world");
+      const confirmationFn = sinon.fake();
+      const errorFn = sinon.fake();
+
+      const foreverCallback = async message => {
+        confirmationFn();
+      };
+
+      await subscription.on("confirmation", foreverCallback);
+
+      await subscription.on("error", err => {
+        errorFn();
+      });
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(confirmationFn.called).to.be.true;
+
+      await new Promise(resolve => setTimeout(resolve, 3000));
+
+      expect(
+        subscription.listenerCount("confirmation"),
```

Comment:
Do an expectation on listenerCount earlier in the test to make sure it was non-zero and changed to zero

Reply:
Should I add that assertion `expect(sub.eventNames()).to.be.empty;` in `beforeEach` hook?
I could create a new `good first issue` but it's a so short task.

Reply:
I've just commited these to PR https://github.com/tasitlabs/TasitSDK/pull/102

Reply:
Thanks! Resolving this.